demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -50,523 +135,127 @@ bool ParseAndCheckCommandLine(int argc, char *argv[]) {
 
-        /** Load extensions for the plugin **/
-        if (!FLAGS_l.empty()) {
-            // CPU(MKLDNN) extensions are loaded as a shared library and passed as a pointer to base extension
-            IExtensionPtr extension_ptr = make_so_pointer<IExtension>(FLAGS_l.c_str());
-            ie.AddExtension(extension_ptr, "CPU");
+        ModelBase* model;
+        if (FLAGS_mt=="ssd")
+        {
+            model = new ModelSSD(FLAGS_m, (float)FLAGS_t, FLAGS_auto_resize, labels);
         }
-        if (!FLAGS_c.empty()) {
-            // clDNN Extensions are loaded from an .xml description and OpenCL kernel files
-            ie.SetConfig({{PluginConfigParams::KEY_CONFIG_FILE, FLAGS_c}}, "GPU");
+        else if (FLAGS_mt=="yolo")
+        {
+            model = new ModelYolo3(FLAGS_m,(float)FLAGS_t, FLAGS_auto_resize, (float)FLAGS_iou_t, labels);
         }
-
-        /** Per layer metrics **/
-        if (FLAGS_pc) {
-            ie.SetConfig({ { PluginConfigParams::KEY_PERF_COUNT, PluginConfigParams::YES } });
+        else
+        {
+            std::cout << "Invalid model type provided: " + FLAGS_mt<<std::endl;
+            return -1;
         }
 
-        std::map<std::string, std::string> userSpecifiedConfig;
-        std::map<std::string, std::string> minLatencyConfig;
-
-        std::set<std::string> devices;
-        for (const std::string& device : parseDevices(FLAGS_d)) {
-            devices.insert(device);
-        }
-        std::map<std::string, unsigned> deviceNstreams = parseValuePerDevice(devices, FLAGS_nstreams);
-        for (auto& device : devices) {
-            if (device == "CPU") {  // CPU supports a few special performance-oriented keys
-                // limit threading for CPU portion of inference
-                if (FLAGS_nthreads != 0)
-                    userSpecifiedConfig.insert({ CONFIG_KEY(CPU_THREADS_NUM), std::to_string(FLAGS_nthreads) });
-
-                if (FLAGS_d.find("MULTI") != std::string::npos
-                    && devices.find("GPU") != devices.end()) {
-                    userSpecifiedConfig.insert({ CONFIG_KEY(CPU_BIND_THREAD), CONFIG_VALUE(NO) });
-                } else {
-                    // pin threads for CPU portion of inference
-                    userSpecifiedConfig.insert({ CONFIG_KEY(CPU_BIND_THREAD), CONFIG_VALUE(YES) });
+        PipelineBase pipeline(model, ConfigFactory::GetUserConfig());
+        Presenter presenter;
+
+        auto startTimePoint = std::chrono::steady_clock::now();
+        while (true){
+            int64_t frameNum;
+            if (pipeline.isReadyToProcess()) {
+                //--- Capturing frame. If previous frame hasn't been inferred yet, reuse it instead of capturing new one
+                if (curr_frame.empty()) {
+                    curr_frame = cap->read();
+                    if (curr_frame.empty())
+                        throw std::logic_error("Can't read an image from the input");
```

Comment:
Empty frame means the input has ended. It's not an error except for the first frame.

Reply:
changed

Reply:
You recreate `frameNum` each iteration and than check it's default initialized value.

Reply:
cap->read may return empty Mat in case of error.